### PR TITLE
Enable dynamic component values

### LINKED ISSUE #1893: Component yaml can not be loaded in k8s if one value is integer and not string
## In what area(s)?

/area operator

## What version of Dapr?
v0.8.0-rc

## Expected Behavior

A component yaml with integer values can be loaded in standalone mode but not if running in kubernetes mode.

Behavior should be same for standalone and kubernetes mode. Either allowing integers or failing with integers.

## Actual Behavior

Standalone is working
Kubernetes throws an error within Operator Service:
```
E0812 06:11:42.076047       1 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.2/tools/cache/reflector.go:105: Failed to list *v1alpha1.Component: v1alpha1.ComponentList.ListMeta: v1.ListMeta.TypeMeta: Kind: Items: []v1alpha1.Component: v1alpha1.Component.Spec: v1alpha1.ComponentSpec.Type: Metadata: []v1alpha1.MetadataItem: v1alpha1.MetadataItem.Value: ReadString: expects " or n, but found 1, error found in #10 byte of ...|,"value":1}],"type":|..., bigger context ...|msQuery","value":""},{"name":"authStyle","value":1}],"type":"middleware.http.oauth2clientcredentials|...
```

## Steps to Reproduce the Problem

Deploy this Component to a dapr k8s
```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: test
spec:
  type: somecomponent
  metadata:
  - name: integer
    value: 1
```
check the logs of the Operator service/pod

### PR DESCRIPTION
**Pull request by yaron2**
Closes #1893.

This PR unifies how `Component` values in YAML are composed for Self Hosted and Kubernetes Modes.

Specifically, it is *no longer required* to wrap values in double quotes when moving component YAMLs to Kubernetes.
I was able to accomplish this using a combination of Kubernetes hidden struct magic and embedded Go types.

Moreover, this was done in a way that is NOT a breaking change. Users are free to use double quoted values or not, in both Kubernetes and Self Hosted modes.

### Kubernetes (Before)

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: statestore
spec:
  type: state.redis
  metadata:
  - name: enableTLS
    value: "true"
 -  name: failoverSeconds
     value: "5"
```

### Kubernetes (After)

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: statestore
spec:
  type: state.redis
  metadata:
  - name: enableTLS
    value: true
 -  name: failoverSeconds
    value: 5
```

I know a lot of people had interest in this, as this also paves the way to https://github.com/dapr/dapr/issues/2030 in the future.

/cc @RichiCoder1 @h2floh @skyao @lynn-orrell

### CODE REVIEW

**Comment by RichiCoder1** [comment]
🥳  This looks great!

File: pkg/apis/components/v1alpha1/types.go [hunk 1]
```diff
@@ -6,6 +6,9 @@
 package v1alpha1
 
 import (
+	"strconv"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
```
Comment (youngbupark) [inline pkg/apis/components/v1alpha1/types.go:11]: Why do we need to import v1beta1? seems like v1 has JSON struct as well.

File: tests/e2e/runtime/runtime_test.go [hunk 2]
```diff
@@ -74,10 +74,10 @@ func TestMain(m *testing.M) {
 			Name:     "runtime-bindings-http",
 			TypeName: "bindings.kafka",
 			MetaData: map[string]string{
-				"brokers":       "dapr-kafka:9092",
-				"topics":        "runtime-bindings-http",
-				"consumerGroup": "group1",
-				"authRequired":  "false",
+				"brokers":       `"dapr-kafka:9092"`,
+				"topics":        `"runtime-bindings-http"`,
+				"consumerGroup": `"group1"`,
```
Comment (youngbupark) [inline tests/e2e/runtime/runtime_test.go:79]: I see all value has quotations. Can we use one of values without quotations? just for reference.

File: pkg/apis/components/v1alpha1/types.go [hunk 3]
```diff
@@ -60,3 +63,19 @@ type ComponentList struct {
 
 	Items []Component `json:"items"`
 }
+
+// DynamicValue is a dynamic value struct for the component.metadata pair value
+type DynamicValue struct {
+	v1beta1.JSON
```
Comment (youngbupark) [inline pkg/apis/components/v1alpha1/types.go:69]: so now user can pass any types of json-object as value, right?

**Review by youngbupark — CHANGES_REQUESTED** [review]

File: pkg/apis/components/v1alpha1/types.go (hunk 3 above)
Comment (mukundansundar) [reply pkg/apis/components/v1alpha1/types.go:69]: Are there any specific validations that must be performed in for the value field?

File: tests/e2e/runtime/runtime_test.go (hunk 2 above)
Comment (yaron2) [reply tests/e2e/runtime/runtime_test.go:79]: The Kubernetes Client requires that for strings are quoted for runtime.JSON values.

File: pkg/apis/components/v1alpha1/types.go (hunk 3 above)
Comment (yaron2) [reply pkg/apis/components/v1alpha1/types.go:69]: > so now user can pass any types of json-object as value, right?

Yep.

Comment (yaron2) [reply pkg/apis/components/v1alpha1/types.go:69]: > Are there any specific validations that must be performed in for the value field?

The Kubernetes Go SDK requires that strings are saved as JSON strings when creating a new `Component` object programmatically. We only do that in our e2e test setup.

File: pkg/apis/components/v1alpha1/types.go (hunk 1 above)
Comment (yaron2) [reply pkg/apis/components/v1alpha1/types.go]: Changed to `v1`.

**Comment by youngbupark** [comment]
Can you remove double-quotation in this state config used by e2e test to simply validate your change in e2e test ?

https://github.com/dapr/dapr/blob/c871b603362e06f9d72d83fee4a6a7d81d50322c/tests/config/dapr_redis_state.yaml#L20

**Review by youngbupark — APPROVED** [review]

**Pull request merged**